File: python/leetcode/333.py
class Solution:
    def largestBSTSubtree(self, root):
        res = 0
        self.dfs(root, res)
        return res

    def dfs(self, root, res):
        if not root: return
        d = self.countBFS(root, -float("Inf"), float("Inf"))
        if d != -1: # the root is bst with left and right node
            res = max(res, d)
            return
        self.dfs(root.left, res)
        self.dfs(root.right, res)

    def countBFS(self, root, mn, mx):
        if not root: return 0
        if root.val < mn or root.val > mx: return -1
        left = self.countBFS(root.left, mn, root.val)
        if left == -1: return -1
        right = self.countBFS(root.right, root.val, mx)
        if right == -1: return -1
        return left + right + 1

File: cpp/leetcode/333.cpp
class Solution {
public:
    int largestBSTSubtree(TreeNode* root) {
        int res = 0;
        dfs(root, res);
        return res;
    }
    void dfs(TreeNode *root, int &res) {
        if (!root) return;
        int d = countBFS(root, INT_MIN, INT_MAX);
        if (d != -1) {
            res = max(res, d);
            return;
        }
        dfs(root->left, res);
        dfs(root->right, res);
    }
    int countBFS(TreeNode *root, int mn, int mx) {
        if (!root) return 0;
        if (root->val < mn || root->val > mx) return -1;
        int left = countBFS(root->left, mn, root->val);
        if (left == -1) return -1;
        int right = countBFS(root->right, root->val, mx);
        if (right == -1) return -1;
        return left + right + 1;
    }
};